idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,165 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
+                sf.append(j)
+        return sf
+
+    def _get_scale_factor_for_product(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i * j)
+        return sf
+
+    def _get_scale_factor_for_division(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i / j)
+        return sf
+
+    def _get_scale_factor_for_power(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        # We will create separate scaling factors for all additive terms
+        # Users can decide how to use these later
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i**j)
+        return sf
+
+    def _get_scale_factor_single_child(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        return child_scale_factors[0]
+
+    def _get_scale_factor_for_unary_function(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        func_name = node.getname()
+        func = getattr(math, func_name)
+        return [func(i) for i in child_scale_factors[0]]
+
+    def _get_scale_factor_expr_if(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 3
+        return child_scale_factors[1] + child_scale_factors[2]
+
+    def _get_scale_factor_no_children(self, node, child_units):
+        assert len(child_units) == 0
+        # may need more checks for dimensionless for other types
+        assert type(node) is IndexTemplate
+        return [1]
+
+    def _get_scale_factor_external_function(self, node, child_units):
+        # TODO: Work out what to do here
+        # Maybe pass child_units into the external function and get the result?
+        # For now, assume we know nothing about scaling
+        return [1]
+
+    node_type_method_map = {
+        EXPR.EqualityExpression: _get_scale_factor_for_sum,
+        EXPR.InequalityExpression: _get_scale_factor_for_sum,
+        EXPR.RangedExpression: _get_scale_factor_for_sum,
+        EXPR.SumExpression: _get_scale_factor_for_sum,
+        EXPR.NPV_SumExpression: _get_scale_factor_for_sum,
+        EXPR.ProductExpression: _get_scale_factor_for_product,
+        EXPR.MonomialTermExpression: _get_scale_factor_for_product,
+        EXPR.NPV_ProductExpression: _get_scale_factor_for_product,
+        EXPR.DivisionExpression: _get_scale_factor_for_division,
+        EXPR.NPV_DivisionExpression: _get_scale_factor_for_division,
+        EXPR.PowExpression: _get_scale_factor_for_power,
+        EXPR.NPV_PowExpression: _get_scale_factor_for_power,
+        EXPR.NegationExpression: _get_scale_factor_single_child,
```

Comment:
One of the shortcomings of this system is that both sums and differences are represented equivalently. Sums of strictly positive quantities can be scaled without user intervention, but when negative numbers are involved [catastrophic cancellation](https://en.wikipedia.org/wiki/Catastrophic_cancellation) can occur, which causes problems for both scaling and numerical methods.

An area in which this regularly occurs in IDAES is with temperature. We deal with absolute temperatures from 300 to 1000K, but temperature differences for heat conduction can be on the order of 1 or 0.1 K.

Reply:
Again, this is why the walker just returns the scaling factor for each term in the sum so the user can go from there.